File: 15-3sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res=[]
        nums.sort()
        for i,n in enumerate(nums):
            if i!=0 and nums[i]==nums[i-1]:
                continue
            l=i+1
            r=len(nums)-1
            while(l<r):
                if n+nums[l]+nums[r]>0:
                    r-=1
                elif n+nums[l]+nums[r]<0:
                    l+=1
                else:
                    res.append([n,nums[l],nums[r]])
                    l+=1
                    while nums[l]==nums[l-1] and l<r:
                        l+=1
                    r-=1
        return res
        # res=[]
        # nums.sort()
        # for i, a in enumerate(nums):
        #     if i > 0 and a == nums[i - 1]:
        #             continue 
        #     l=i+1
        #     r=len(nums)-1
        #     while(l<r):
        #         if nums[l]+nums[r]+a>0:
        #             r-=1
        #         elif nums[l]+nums[r]+a<0:
        #             l+=1
        #         else:
        #             res.append([a,nums[l],nums[r]])
        #             l += 1
        #             r -= 1
        #             while nums[l] == nums[l - 1] and l < r:
        #                 l += 1
        # return res
            

File: 15-3sum/3sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        Array.Sort(nums);
        var res = new List<IList<int>>();
        for (int i=0;i<nums.Length;i++){
            if(i!=0 && nums[i-1]==nums[i]){
                continue;
            }
            int st=i+1;
            int end=nums.Length-1;
            while(st<end){
                if(nums[st]+nums[end]+nums[i]==0){
                    res.Add(new List<int>{nums[st],nums[end],nums[i]});
                    st++;
                    end--;
                    while(nums[st]==nums[st-1] && st<end){
                        st++;
                    }
                }else if(nums[st]+nums[end]+nums[i]<0){
                    st++;
                }else{
                    end--;
                }
            }
        }
        return res;
    }
}